### ISSUE TITLE
[xray] Actor resource only gets released when python GC happens

### ISSUE THREAD

**Issue by guoyuhong** [ray-project/ray#2641]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux/MacOs
- **Ray installed from (source or binary)**: Source
- **Ray version**: 0.5
- **Python version**: 3.6/2.7
- **Exact command to reproduce**:

### Describe the problem

Current actor resource management is [a lifetime resource](https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L1026-L1030), which means once an actor is created, it will occupy the resource until this actor is [disconnected](DisconnectClient). However, this disconnect message will be sent once the instance of an `ActorHandle` is garbage collected ([This code snippet](https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L863-L875) shows that only in `__del__` of `ActorHandle` the `DisconnectClient` message will be send.). And there is no warning when this case happens.

In my understanding, precious resource should not be determined by garbage collection behavior. I think we can provide a `release` function like the close function in file handle of python. At the time we can also provide a `with` statement for actor to release the resource. 

### Source code / logs

https://github.com/ray-project/ray/issues/2588

**Comment by ericl** [comment]
Also related: https://github.com/ray-project/ray/issues/1516

**Comment by zhijunfu** [comment]
Agreed. I encountered the same problem as well during perf testing.

**Comment by robertnishihara** [comment]
@guoyuhong doesn't the GC happen as soon as the actor handle object goes out of scope? (assuming there are no cycles).

Anyway, I agree that we could have some sort of "release" functionality. Currently we work around this by doing something like `actor.__ray_terminate__.remote()`, although that should be considered an implementation detail and not part of the API.

**Comment by guoyuhong** [comment]
From the actor behavior, it is not called. That is why there is no enough resource for later actors in https://github.com/ray-project/ray/issues/2588 .

**Comment by robertnishihara** [comment]
I believe #2588 is a somewhat different issue. In that case, as you point out, the actor handles are not going out of scope. However, we do not want to GC the actors in that application, we actually want to keep all of the actors around.

**Comment by ericl** [comment]
Auto-closing stale issue.
